File: Trabajo_8/PY/Ejercicios/Ejercicio_2.py
from abc import ABC, abstractmethod
import time


class Reproductor(ABC):
    @abstractmethod
    def reproducir(self):
        pass
    @abstractmethod
    def pausar(self):
        pass
    @abstractmethod
    def detener(self):
        pass
    @abstractmethod
    def encender(self):
        pass

    @abstractmethod
    def apagar(self):
        pass

class ReproductorMP3(Reproductor):
    def __init__(self):
        # modo (pausado-reproduciento)
        self.estado = False
        # representa si esta apagado o encendido
        self.modo = False

    def reproducir(self):
        if self.modo == True:
            if self.estado == True:
                print("Reanunando reproduccion")
            else:
                self.modo = True
                print(f"Reproduciendo archivo mp3")
        else:
            print("Reproductor apagado")

    def pausar(self):
        if self.modo == True:
            if self.estado == False:
                self.estado = True
                print("Pausando reproduccion")
            else:
                print("Archivo mp3 ya en pausa")
        else:
            print("Reproductor apagado")

    def detener(self):
        if self.modo == True:
            self.estado = False
            print("Deteniendo Reproduccion de musica")
        else:
            print("Dispositivo apagado")

    def encender(self):
        if self.modo == False:
            self.modo = True
            print("Encendiendo MP3...")

    def apagar(self):
        if self.modo == True:
            self.modo = False
            self.estado = False
            print("Apagando Reproductor")

def menu():
    print(" ***** Menu ***** ")
    print("1. Encender Reproductor")
    print("2. Reproducir archivo Mp3")
    print("3. Pausar archivo Mp3")
    print("4. Detener archivo Mp3 ")
    print("5. Apagar Reproductor")

def menu_acciones():
    print("Menu Acciones")
    print("1. Reproducir archivo Mp3")
    print("2. Pausar archivo Mp3")
    print("3. Detener archivo Mp3 ")

if __name__ == "__main__":
    mp3 = ReproductorMP3()
    while True:
        menu()
        op = int(input("Ingrese una opcion: "))
        if op == 1:
            mp3.encender()


        elif op == 2:
            mp3.reproducir()


        elif op == 3:
            mp3.pausar()


        elif op == 4:
            mp3.detener()


        elif op == 5:
            mp3.apagar()
            break
        else:
            print("Opcion invalida")
        print("\n")
        time.sleep(1.5)

File: Trabajo_8/CPP/Ejercicios/Ejercicio_2.cpp
#include <iostream>
using namespace std;

//Interfaz
class Reproductor{
    //En este caso todos van a manipular el objeto, por lo que ninguno tiene el const
    virtual void reproducir() = 0;
    virtual void pausar() = 0;
    virtual void detener() = 0;
    virtual void iniciar() = 0;
};


//Clase Derivada
class ReproductorMP3 : public Reproductor{
private:
    bool estado = 0;
    bool modo = 0;

public:
    void reproducir() override{
        if (estado == 1){
            cout << "La musica ya se encuentra en reproduccion..." << endl;
        } else{
            this -> estado = 1;
            cout << "Reproduciendo..." << endl;
        }
    }

    void pausar() override{
        if (estado == 0){
            cout << "La música ya esta pausada..." << endl;
        } else{
            this -> estado = 0;
            cout << "Pausando..." << endl;
        }
    }

    void iniciar() override{
        this -> modo = 1;
        cout << "Bienvenido..." << endl;
    }

    void detener() override{
        this -> modo = 0;
        cout << "Hasta pronto..." << endl;
    }

    friend ostream& operator <<(ostream& os, ReproductorMP3 R_01){
        if (R_01.estado == 0){
            os << "Estado del reproductor --> Pausado..." << endl;
            return os;
        } else {
            os << "Estado del reproductor --> Reproduciendo..." << endl;
            return os;
        }
    }

    bool &getmodo(){
        return modo;
    }
};

void menu_acciones_on_off(){
    cout << "1 - Encender Reproductor" << endl;
    cout << "0 - Terminar Programa" << endl;
}

void menu_acciones_rep(){
    cout << "1 - Reproducir" << endl;
    cout << "2 - Pausar" << endl;
    cout << "3 - Mostrar Estado" << endl;
    cout << "0 - Detener" << endl;
}


int verificar(int end){
    while (true){
        int num;
        cout << "-->";
        cin >> num;
        for (int i = 0; i < end; i++){
            if (i == num){
                return num;
            }
        }
    }
}



int main(){
    ReproductorMP3 Rep_01;
    bool ejecutor = true;

    while(ejecutor){

        int eleccion;
        menu_acciones_on_off();
        eleccion = verificar(2);
        if (eleccion == 1){
            Rep_01.iniciar();
        }else{
            return 0;
        }

        while (Rep_01.getmodo())
        {
            int election;
            menu_acciones_rep();
            election = verificar(4);
            if (election == 1){
                Rep_01.reproducir();
            } else if (election == 2){
                Rep_01.pausar();
            } else if (election == 3){
                cout << Rep_01;
            } else{
                Rep_01.detener();
            }
        }
    }
}
